dvc/objects/external.py

```diff
@@ -0,0 +1,107 @@
+import logging
+from typing import TYPE_CHECKING, Optional
+
+from dvc.path_info import PathInfo
+
+from .errors import ObjectFormatError
+from .file import HashFile
+
+if TYPE_CHECKING:
+    from dvc.hash_info import HashInfo
+    from dvc.types import DvcPath
+
+    from .db.base import ObjectDB
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalRepoFile(HashFile):
+    """Lazy erepo object.
+
+    Will be staged in def_odb when object hash is needed.
+    """
+
+    def __init__(
+        self,
+        def_odb: "ObjectDB",
+        repo_url: str,
+        repo_rev: Optional[str],
+        path_info: "DvcPath",
+        name: Optional[str] = None,
+    ):  # pylint: disable=super-init-not-called
+        self.def_odb = def_odb
+        self.repo_url = repo_url
+        self.repo_rev = repo_rev
+        self._resolved_rev = None
+        self.path_info = path_info
+        self._hash_info: Optional["HashInfo"] = None
+        self.name = name
+
+    def __str__(self):
+        return (
+            f"external object {self.repo_url}: {self.path_info} "
+            f"@ {self.repo_rev}"
+        )
+
+    def __bool__(self):
+        return bool(self._hash_info)
+
+    def __eq__(self, other):
+        if not isinstance(other, ExternalRepoFile):
+            return False
+        return (
+            self.repo_url == other.repo_rev
+            and self.path_info == other.path_info
+        )
+
+    def __hash__(self):
+        return hash((self.repo_url, self.repo_rev, self.path_info))
+
+    def _make_repo(self, **kwargs):
+        from dvc.external_repo import external_repo
+        from dvc.scm.base import CloneError
+
+        try:
+            return external_repo(self.repo_url, rev=self.repo_rev, **kwargs)
+        except CloneError as exc:
+            raise ObjectFormatError(
+                f"Could not clone external obj repo '{self.repo_url}'"
+            ) from exc
+
+    def get_staged(self, odb: Optional["ObjectDB"] = None):
```

Comment:
It feels a bit off to have an object stage itself, but in this case external repo deps have to be staged before we can compute the hash_info